File: app/models/sections/ExamSectionQuestion.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
// SPDX-FileCopyrightText: 2024. The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package models.sections;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.databind.JsonNode;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Transient;
import java.math.BigDecimal;
import java.math.MathContext;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.TreeMap;
import javax.annotation.Nonnull;
import models.base.OwnedModel;
import models.questions.ClozeTestAnswer;
import models.questions.EssayAnswer;
import models.questions.MultipleChoiceOption;
import models.questions.Question;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.BeanUtils;
import play.mvc.Result;

@Entity
public class ExamSectionQuestion extends OwnedModel implements Comparable<ExamSectionQuestion>, Sortable, Scorable {

    private final Logger logger = LoggerFactory.getLogger(ExamSectionQuestion.class);

    @ManyToOne
    @JoinColumn(name = "exam_section_id")
    @JsonBackReference
    private ExamSection examSection;

    @ManyToOne(cascade = CascadeType.PERSIST)
    @JoinColumn(name = "question_id")
    private Question question;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "examSectionQuestion")
    private Set<ExamSectionQuestionOption> options;

    @Column
    private int sequenceNumber;

    @Column
    private String answerInstructions;

    @Column
    private Double maxScore;

    @Column
    private Double forcedScore;

    @Transient
    private Double derivedMaxScore;

    @Transient
    private Double derivedAssessedScore;

    @Transient
    private Double derivedMinScore;

    @Column
    private Question.EvaluationType evaluationType;

    @OneToOne(cascade = CascadeType.ALL)
    private EssayAnswer essayAnswer;

    @OneToOne(cascade = CascadeType.ALL)
    private ClozeTestAnswer clozeTestAnswer;

    @Column
    private String evaluationCriteria;

    @Column
    private Integer expectedWordCount;

    public ExamSection getExamSection() {
        return examSection;
    }

    public void setExamSection(ExamSection examSection) {
        this.examSection = examSection;
    }

    public Question getQuestion() {
        return question;
    }

    public void setQuestion(Question question) {
        this.question = question;
    }

    public Set<ExamSectionQuestionOption> getOptions() {
        return options;
    }

    public void setOptions(Set<ExamSectionQuestionOption> options) {
        this.options = options;
    }

    public int getSequenceNumber() {
        return sequenceNumber;
    }

    public void setSequenceNumber(int sequenceNumber) {
        this.sequenceNumber = sequenceNumber;
    }

    public Double getMaxScore() {
        return maxScore;
    }

    public void setMaxScore(Double maxScore) {
        this.maxScore = maxScore;
    }

    public Double getForcedScore() {
        return forcedScore;
    }

    public void setForcedScore(Double forcedScore) {
        this.forcedScore = forcedScore;
    }

    public Double getDerivedMaxScore() {
        return derivedMaxScore;
    }

    public void setDerivedMaxScore() {
        this.derivedMaxScore = getMaxAssessedScore();
    }

    public void setDerivedAssessedScore() {
        this.derivedAssessedScore = getAssessedScore();
    }

    public Double getDerivedMinScore() {
        return derivedMinScore;
    }

    public void setDerivedMinScore() {
        this.derivedMinScore = getMinScore();
    }

    public String getAnswerInstructions() {
        return answerInstructions;
    }

    public void setAnswerInstructions(String answerInstructions) {
        this.answerInstructions = answerInstructions;
    }

    public Question.EvaluationType getEvaluationType() {
        return evaluationType;
    }

    public void setEvaluationType(Question.EvaluationType evaluationType) {
        this.evaluationType = evaluationType;
    }

    public EssayAnswer getEssayAnswer() {
        return essayAnswer;
    }

    public void setEssayAnswer(EssayAnswer essayAnswer) {
        this.essayAnswer = essayAnswer;
    }

    public ClozeTestAnswer getClozeTestAnswer() {
        return clozeTestAnswer;
    }

    public void setClozeTestAnswer(ClozeTestAnswer clozeTestAnswer) {
        this.clozeTestAnswer = clozeTestAnswer;
    }

    public String getEvaluationCriteria() {
        return evaluationCriteria;
    }

    public void setEvaluationCriteria(String evaluationCriteria) {
        this.evaluationCriteria = evaluationCriteria;
    }

    public Integer getExpectedWordCount() {
        return expectedWordCount;
    }

    public void setExpectedWordCount(Integer expectedWordCount) {
        this.expectedWordCount = expectedWordCount;
    }

    ExamSectionQuestion copyWithAnswers(Boolean hasParent) {
        ExamSectionQuestion esqCopy = new ExamSectionQuestion();
        BeanUtils.copyProperties(this, esqCopy, "id", "options", "essayAnswer", "clozeTestAnswer");
        // This is a little bit tricky. Need to map the original question options with copied ones, so they can be
        // associated with both question and exam section question options :)

        Map<Long, MultipleChoiceOption> optionMap;

        if (question.getType() == Question.Type.ClaimChoiceQuestion) {
            optionMap = new TreeMap<>();
        } else {
            optionMap = new HashMap<>();
        }

        Question blueprint = question.copy(optionMap, hasParent);
        if (hasParent) {
            blueprint.setParent(question);
        }
        blueprint.save();
        options.forEach(option -> {
            Optional<MultipleChoiceOption> parentOption = Optional.ofNullable(option.getOption()).filter(
                opt -> opt.getId() != null
            );
            if (parentOption.isPresent()) {
                MultipleChoiceOption optionCopy = optionMap.get(parentOption.get().getId());
                optionCopy.setQuestion(blueprint);
                optionCopy.save();
                ExamSectionQuestionOption esqoCopy = option.copyWithAnswer();
                esqoCopy.setOption(optionCopy);
                esqCopy.getOptions().add(esqoCopy);
            } else {
                logger.error("Failed to copy a multi-choice question option!");
                throw new RuntimeException();
            }
        });

        esqCopy.setQuestion(blueprint);
        // Essay Answer
        if (essayAnswer != null) {
            esqCopy.setEssayAnswer(essayAnswer.copy());
        }
        if (clozeTestAnswer != null) {
            esqCopy.setClozeTestAnswer(clozeTestAnswer.copy());
        }
        return esqCopy;
    }

    ExamSectionQuestion copy(boolean preserveOriginalQuestion, boolean setParent) {
        ExamSectionQuestion esqCopy = new ExamSectionQuestion();
        BeanUtils.copyProperties(this, esqCopy, "id", "options", "creator", "modifier");
        Question blueprint;
        if (preserveOriginalQuestion) {
            // Use the existing question references, no copying
            blueprint = question;
            options.forEach(o -> esqCopy.getOptions().add(o.copy()));
        } else {
            // This is a little bit tricky. Need to map the original question options with copied ones, so they can be
            // associated with both question and exam section question options :)
            Map<Long, MultipleChoiceOption> optionMap;

            if (question.getType() == Question.Type.ClaimChoiceQuestion) {
                optionMap = new TreeMap<>();
            } else {
                optionMap = new HashMap<>();
            }

            blueprint = question.copy(optionMap, setParent);
            if (setParent) {
                blueprint.setParent(question);
            }
            blueprint.save();
            optionMap.forEach((k, optionCopy) -> {
                optionCopy.setQuestion(blueprint);
                optionCopy.save();
                options
                    .stream()
                    .filter(o -> o.getOption().getId().equals(k))
                    .findFirst()
                    .ifPresentOrElse(
                        esqo -> {
                            ExamSectionQuestionOption esqoCopy = esqo.copy();
                            esqoCopy.setOption(optionCopy);
                            esqCopy.getOptions().add(esqoCopy);
                        },
                        () -> {
                            logger.error("Failed to copy a multi-choice question option!");
                            throw new RuntimeException();
                        }
                    );
            });
        }
        esqCopy.setQuestion(blueprint);
        return esqCopy;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof ExamSectionQuestion other)) {
            return false;
        }
        return new EqualsBuilder().append(examSection, other.examSection).append(question, other.question).build();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder().append(examSection).append(question).build();
    }

    @Override
    public int compareTo(@Nonnull ExamSectionQuestion o) {
        return sequenceNumber - o.sequenceNumber;
    }

    @Override
    public Integer getOrdinal() {
        return sequenceNumber;
    }

    @Override
    public void setOrdinal(Integer ordinal) {
        sequenceNumber = ordinal;
    }

    @Override
    public Optional<Result> getValidationResult(JsonNode node) {
        return question.getValidationResult(node);
    }

    @Override
    public Double getAssessedScore() {
        switch (question.getType()) {
            case EssayQuestion -> {
                if (evaluationType == Question.EvaluationType.Points) {
                    return essayAnswer == null || essayAnswer.getEvaluatedScore() == null
                        ? 0
                        : essayAnswer.getEvaluatedScore();
                }
            }
            case MultipleChoiceQuestion -> {
                if (forcedScore != null) {
                    return forcedScore;
                }
                Optional<ExamSectionQuestionOption> o = options
                    .stream()
                    .filter(ExamSectionQuestionOption::isAnswered)
                    .findFirst();
                if (o.isPresent()) {
                    return o.get().getOption().isCorrectOption() ? maxScore : 0.0;
                }
            }
            case WeightedMultipleChoiceQuestion -> {
                if (forcedScore != null) {
                    return forcedScore;
                }
                Double evaluation = options
                    .stream()
                    .filter(esq -> esq.isAnswered() && esq.getScore() != null)
                    .map(ExamSectionQuestionOption::getScore)
                    .reduce(0.0, Double::sum);
                // ATM minimum score is zero
                return Math.max(0.0, evaluation);
            }
            case ClozeTestQuestion -> {
                if (forcedScore != null) {
                    return forcedScore;
                }
                // sanity check
                if (clozeTestAnswer == null) {
                    return 0.0;
                }
                ClozeTestAnswer.Score score = clozeTestAnswer.calculateScore(this);
                int correct = score.getCorrectAnswers();
                int incorrect = score.getIncorrectAnswers();
                if (correct + incorrect == 0) {
                    return 0.0;
                }
                DecimalFormat df = new DecimalFormat("#.##", new DecimalFormatSymbols(Locale.US));
                double value = (correct * maxScore) / (correct + incorrect);
                return Double.valueOf(df.format(value));
            }
            case ClaimChoiceQuestion -> {
                if (forcedScore != null) {
                    return forcedScore;
                }
                Optional<ExamSectionQuestionOption> answeredOption = options
                    .stream()
                    .filter(ExamSectionQuestionOption::isAnswered)
                    .findFirst();
                if (answeredOption.isPresent()) {
                    return answeredOption.get().getScore();
                }
                return 0.0;
            }
        }
        return 0.0;
    }

    @Override
    public Double getMaxAssessedScore() {
        switch (question.getType()) {
            case EssayQuestion -> {
                if (evaluationType == Question.EvaluationType.Points) {
                    return maxScore == null ? 0 : maxScore;
                }
            }
            case MultipleChoiceQuestion, ClozeTestQuestion -> {
                return maxScore == null ? 0 : maxScore;
            }
            case WeightedMultipleChoiceQuestion -> {
                return options
                    .stream()
                    .map(ExamSectionQuestionOption::getScore)
                    .filter(score -> score != null && score > 0)
                    .reduce(0.0, Double::sum);
            }
            case ClaimChoiceQuestion -> {
                return options
                    .stream()
                    .map(ExamSectionQuestionOption::getScore)
                    .filter(Objects::nonNull)
                    .max(Comparator.comparing(Double::valueOf))
                    .orElse(0.0);
            }
        }
        return 0.0;
    }

    public Double getMinScore() {
        if (question.getType() == Question.Type.WeightedMultipleChoiceQuestion) {
            return options
                .stream()
                .map(ExamSectionQuestionOption::getScore)
                .filter(score -> score != null && score < 0)
                .reduce(0.0, Double::sum);
        } else if (question.getType() == Question.Type.ClaimChoiceQuestion) {
            return options
                .stream()
                .map(ExamSectionQuestionOption::getScore)
                .filter(Objects::nonNull)
                .min(Comparator.comparing(Double::valueOf))
                .orElse(0.0);
        }
        return 0.0;
    }

    @Override
    public boolean isRejected() {
        return (
            question.getType() == Question.Type.EssayQuestion &&
            evaluationType == Question.EvaluationType.Selection &&
            essayAnswer != null &&
            essayAnswer.getEvaluatedScore() != null &&
            essayAnswer.getEvaluatedScore() == 0
        );
    }

    @Override
    public boolean isApproved() {
        return (
            question.getType() == Question.Type.EssayQuestion &&
            evaluationType == Question.EvaluationType.Selection &&
            essayAnswer != null &&
            essayAnswer.getEvaluatedScore() != null &&
            essayAnswer.getEvaluatedScore() == 1
        );
    }

    /**
     * Adds new answer option.
     * If question type equals WeightedMultiChoiceQuestion, recalculates scores for old options so that max assessed
     * score won't change.
     *
     * @param option New option to add.
     */
    public void addOption(ExamSectionQuestionOption option, boolean preserveScores) {
        if (question.getType() == Question.Type.ClaimChoiceQuestion) return;

        if (
            question.getType() != Question.Type.WeightedMultipleChoiceQuestion ||
            option.getScore() == null ||
            preserveScores
        ) {
            options.add(option);
            return;
        }

        if (option.getScore() > 0) {
            List<ExamSectionQuestionOption> opts = options
                .stream()
                .filter(o -> o.getScore() != null && o.getScore() > 0)
                .toList();
            BigDecimal delta = calculateOptionScores(option.getScore(), opts);
            option.setScore(BigDecimal.valueOf(option.getScore()).add(delta).doubleValue());
        } else if (option.getScore() < 0) {
            List<ExamSectionQuestionOption> opts = options
                .stream()
                .filter(o -> o.getScore() != null && o.getScore() < 0)
                .toList();
            BigDecimal delta = calculateOptionScores(option.getScore(), opts);
            option.setScore(BigDecimal.valueOf(option.getScore()).add(delta).doubleValue());
        }
        options.add(option);
    }

    private void initOptionScore(Double score, List<ExamSectionQuestionOption> options) {
        BigDecimal delta = calculateOptionScores(score * -1, options);
        if (!options.isEmpty()) {
            ExamSectionQuestionOption first = options.getFirst();
            first.setScore(BigDecimal.valueOf(first.getScore()).add(delta).doubleValue());
        }
    }

    public void removeOption(MultipleChoiceOption option, boolean preserveScores) {
        if (question.getType() == Question.Type.ClaimChoiceQuestion) return;

        ExamSectionQuestionOption esqo = options
            .stream()
            .filter(o -> option.equals(o.getOption()))
            .findFirst()
            .orElse(null);
        if (esqo == null) {
            return;
        }

        Double score = esqo.getScore();
        options.remove(esqo);
        if (question.getType() != Question.Type.WeightedMultipleChoiceQuestion || score == null || preserveScores) {
            return;
        }

        if (score > 0) {
            List<ExamSectionQuestionOption> opts = options
                .stream()
                .filter(o -> o.getScore() != null && o.getScore() > 0)
                .toList();
            initOptionScore(score, opts);
        } else if (score < 0) {
            List<ExamSectionQuestionOption> opts = options
                .stream()
                .filter(o -> o.getScore() != null && o.getScore() < 0)
                .toList();
            initOptionScore(score, opts);
        }
    }

    private BigDecimal calculateOptionScores(double optionScore, List<ExamSectionQuestionOption> opts) {
        BigDecimal oldSum = new BigDecimal(0);
        BigDecimal sum = new BigDecimal(0);
        for (ExamSectionQuestionOption o : opts) {
            oldSum = oldSum.add(BigDecimal.valueOf(o.getScore()));
            o.setScore(roundToTwoDigits(o.getScore() - optionScore / opts.size()));
            sum = sum.add(BigDecimal.valueOf(o.getScore()));
        }
        BigDecimal subtracted = oldSum.subtract(BigDecimal.valueOf(optionScore).add(sum));
        return subtracted.round(new MathContext(1));
    }

    private Double roundToTwoDigits(double score) {
        return Math.round(score * 100) / 100d;
    }
}

File: app/models/sections/ExamSectionQuestion.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package models.sections

import com.fasterxml.jackson.annotation.{JsonBackReference, JsonProperty}
import features.exam.copy.ExamCopyContext
import jakarta.persistence.*
import models.base.OwnedModel
import models.questions.*
import play.api.Logging

import scala.compiletime.uninitialized
import scala.jdk.CollectionConverters.*
import scala.util.Random

@Entity
class ExamSectionQuestion extends OwnedModel with Ordered[ExamSectionQuestion] with Sortable
    with Scorable with Logging:

  @ManyToOne
  @JoinColumn(name = "exam_section_id")
  @JsonBackReference
  var examSection: ExamSection = uninitialized

  @ManyToOne(cascade = Array(CascadeType.PERSIST))
  @JoinColumn(name = "question_id")
  var question: Question = uninitialized

  @OneToMany(
    cascade = Array(CascadeType.ALL),
    orphanRemoval = true,
    mappedBy = "examSectionQuestion"
  )
  @OrderBy("option.id") // TODO: is this needed somewhere?
  var options: java.util.List[ExamSectionQuestionOption] = uninitialized

  @OneToOne(cascade = Array(CascadeType.ALL))
  @JsonProperty
  var essayAnswer: EssayAnswer = uninitialized

  @OneToOne(cascade = Array(CascadeType.ALL))
  @JsonProperty
  var clozeTestAnswer: ClozeTestAnswer = uninitialized

  var sequenceNumber: Int                    = 0
  var answerInstructions: String             = uninitialized
  var maxScore: java.lang.Double             = uninitialized
  var negativeScoreAllowed: Boolean          = false
  var optionShufflingOn: Boolean             = false
  var forcedScore: java.lang.Double          = uninitialized
  var evaluationType: QuestionEvaluationType = uninitialized
  var evaluationCriteria: String             = uninitialized
  var expectedWordCount: Integer             = uninitialized

  @Transient var derivedMaxScore: Double      = uninitialized
  @Transient var derivedAssessedScore: Double = uninitialized
  @Transient var derivedMinScore: Double      = uninitialized
  def setDerivedMaxScore(): Unit              = derivedMaxScore = getMaxAssessedScore
  def setDerivedAssessedScore(): Unit         = derivedAssessedScore = getAssessedScore
  def setDerivedMinScore(): Unit              = derivedMinScore = getMinScore

  override def getOrdinal: Integer          = sequenceNumber
  override def setOrdinal(o: Integer): Unit = sequenceNumber = o

  override def getAssessedScore: Double =
    // Helper to handle the recurring forcedScore logic safely
    def withForcedScore(fallback: => Double): Double =
      Option(forcedScore).filter(_ != 0).map(_.toDouble).getOrElse(fallback)

    question.`type` match
      case QuestionType.EssayQuestion =>
        val score = if evaluationType == QuestionEvaluationType.Points then
          Option(essayAnswer).map(_.evaluatedScore.toDouble).getOrElse(0.0)
        else 0.0
        if score == 0 then 0.0 else score
      case QuestionType.MultipleChoiceQuestion =>
        withForcedScore {
          options.asScala
            .find(_.answered)
            .filter(_.option.correctOption)
            .map(_ => maxScore.toDouble) // Force to primitive
            .getOrElse(0.0)
        }
      case QuestionType.WeightedMultipleChoiceQuestion =>
        withForcedScore {
          val eval = options.asScala.filter(_.answered).map(_.score.toDouble).sum
          if negativeScoreAllowed then eval else math.max(0.0, eval)
        }
      case QuestionType.ClozeTestQuestion =>
        withForcedScore {
          Option(clozeTestAnswer).map(_.calculateScore(this)) match
            case Some(s) if (s.correctAnswers + s.incorrectAnswers) > 0 =>
              val total = (s.correctAnswers + s.incorrectAnswers).toDouble
              BigDecimal(s.correctAnswers.toDouble * maxScore.toDouble / total)
                .setScale(2, BigDecimal.RoundingMode.HALF_UP)
                .toDouble
            case _ => 0.0
        }
      case QuestionType.ClaimChoiceQuestion =>
        withForcedScore { options.asScala.find(_.answered).map(_.score.toDouble).getOrElse(0.0) }
      case null => 0.0

  override def getMaxAssessedScore: Double =
    question.`type` match
      case QuestionType.EssayQuestion =>
        if evaluationType == QuestionEvaluationType.Points then
          if maxScore == 0 then 0 else maxScore
        else 0.0
      case QuestionType.MultipleChoiceQuestion | QuestionType.ClozeTestQuestion =>
        if maxScore == 0 then 0 else maxScore
      case QuestionType.WeightedMultipleChoiceQuestion =>
        options.asScala.filter(o => o.isLegitScore(negativeScoreAllowed) && o.score >= 0).map(
          _.score.toDouble
        ).sum
      case QuestionType.ClaimChoiceQuestion =>
        options.asScala.map(_.score.toDouble).filter(_ != 0).maxOption.getOrElse(0.0)
      case null => 0.0 // This should never happen, java enum interop stuff

  def getMinScore: Double =
    question.`type` match
      case QuestionType.WeightedMultipleChoiceQuestion =>
        options.asScala.filter(o => o.isLegitScore(negativeScoreAllowed) && o.score <= 0).map(
          _.score.toDouble
        ).sum
      case QuestionType.ClaimChoiceQuestion =>
        options.asScala.map(_.score.toDouble).filter(_ != 0).minOption.getOrElse(0.0)
      case _ => 0.0

  override def isRejected: Boolean =
    question.`type` == QuestionType.EssayQuestion &&
      evaluationType == QuestionEvaluationType.Selection &&
      essayAnswer != null && essayAnswer.evaluatedScore == 0

  override def isApproved: Boolean =
    question.`type` == QuestionType.EssayQuestion &&
      evaluationType == QuestionEvaluationType.Selection &&
      essayAnswer != null && essayAnswer.evaluatedScore == 1

  def copy(context: ExamCopyContext): ExamSectionQuestion =
    val esqCopy = new ExamSectionQuestion
    if context.shouldCopyAnswers then
      copyScalarFields(esqCopy)
      esqCopy.question = question
      esqCopy.evaluationType = evaluationType
      copyQuestionWithAnswers(esqCopy, context)
      if essayAnswer != null then esqCopy.essayAnswer = essayAnswer.copy()
      if clozeTestAnswer != null then esqCopy.clozeTestAnswer = clozeTestAnswer.copy()
    else
      copyScalarFields(esqCopy)
      esqCopy.evaluationType = evaluationType
      if context.isStudentExam then
        copyQuestionWithOptions(esqCopy, context)
        if optionShufflingOn && question.`type` != QuestionType.ClaimChoiceQuestion then
          esqCopy.shuffleOptions()
      else
        esqCopy.question = question
        esqCopy.creator = creator
        esqCopy.created = created
        options.asScala.foreach(o => esqCopy.options.add(o.copy()))
    esqCopy

  private def copyScalarFields(dest: ExamSectionQuestion): Unit =
    dest.sequenceNumber = sequenceNumber
    dest.answerInstructions = answerInstructions
    dest.maxScore = maxScore
    dest.negativeScoreAllowed = negativeScoreAllowed
    dest.optionShufflingOn = optionShufflingOn
    dest.forcedScore = forcedScore
    dest.evaluationCriteria = evaluationCriteria
    dest.expectedWordCount = expectedWordCount

  private def copyQuestionWithAnswers(
      esqCopy: ExamSectionQuestion,
      context: ExamCopyContext
  ): Unit =
    val (blueprint, optionMap) = question.copyWithOptions(context.shouldSetParent)
    if context.shouldSetParent then blueprint.parent = question
    blueprint.save()
    persistQuestionOwners(blueprint, context)
    options.asScala.foreach { opt =>
      Option(opt.option).filter(_.id != 0) match
        case Some(parentOpt) =>
          optionMap.get(parentOpt.id) match
            case Some(optCopy) =>
              optCopy.question = blueprint
              optCopy.save()
              val esqoCopy = opt.copyWithAnswer()
              esqoCopy.option = optCopy
              esqCopy.options.add(esqoCopy)
            case None =>
              logger.error("Failed to copy a multi-choice question option!")
              throw new RuntimeException()
        case None =>
          logger.error("Failed to copy a multi-choice question option!")
          throw new RuntimeException()
    }
    esqCopy.question = blueprint

  private def copyQuestionWithOptions(
      esqCopy: ExamSectionQuestion,
      context: ExamCopyContext
  ): Unit =
    val (blueprint, optionMap) = question.copyWithOptions(context.shouldSetParent)
    if context.shouldSetParent then blueprint.parent = question
    blueprint.save()
    persistQuestionOwners(blueprint, context)
    optionMap.foreach { (k, optCopy) =>
      optCopy.question = blueprint
      optCopy.save()
      options.asScala.find(_.option.id == k) match
        case Some(esqo) =>
          val esqoCopy = esqo.copy()
          esqoCopy.option = optCopy
          esqCopy.options.add(esqoCopy)
        case None =>
          logger.error("Failed to copy a multi-choice question option!")
          throw new RuntimeException()
    }
    esqCopy.question = blueprint

  private def persistQuestionOwners(blueprint: Question, context: ExamCopyContext): Unit =
    if context.shouldSetParent && !blueprint.questionOwners.isEmpty then
      blueprint.questionOwners = new java.util.HashSet(blueprint.questionOwners)
      blueprint.save()

  private def shuffleOptions(): Unit =
    // need to wrap with a java list for interop with ebean/gson
    options = new java.util.ArrayList(Random.shuffle(options.asScala.toSeq).asJava)

  def addOption(option: ExamSectionQuestionOption, preserveScores: Boolean): Unit =
    if question.`type` != QuestionType.ClaimChoiceQuestion then
      if question.`type` != QuestionType.WeightedMultipleChoiceQuestion || option.score == 0 || preserveScores
      then options.add(option)
      else
        if option.score > 0 then
          val opts  = options.asScala.filter(o => o.score != 0 && o.score > 0).toSeq
          val delta = calculateOptionScores(option.score, opts)
          option.score = option.score.toDouble + delta
        else if option.score < 0 then
          val opts  = options.asScala.filter(o => o.score != 0 && o.score < 0).toSeq
          val delta = calculateOptionScores(option.score, opts)
          option.score = option.score.toDouble + delta
        options.add(option)

  def removeOption(option: MultipleChoiceOption, preserveScores: Boolean): Unit =
    if question.`type` != QuestionType.ClaimChoiceQuestion then
      options.asScala.find(o => option == o.option).foreach { esqo =>
        val score = esqo.score
        options.remove(esqo)
        if question.`type` == QuestionType.WeightedMultipleChoiceQuestion && score != 0 && !preserveScores
        then
          if score > 0 then
            val opts = options.asScala.filter(o => o.score != 0 && o.score > 0).toSeq
            initOptionScore(score, opts)
          else if score < 0 then
            val opts = options.asScala.filter(o => o.score != 0 && o.score < 0).toSeq
            initOptionScore(score, opts)
      }

  private def initOptionScore(score: Double, opts: Seq[ExamSectionQuestionOption]): Unit =
    val delta = calculateOptionScores(score * -1, opts)
    opts.headOption.foreach(first =>
      first.score = first.score.toDouble + delta
    )

  private def calculateOptionScores(
      optionScore: Double,
      opts: Seq[ExamSectionQuestionOption]
  ): Double =
    var oldSum = BigDecimal(0)
    var sum    = BigDecimal(0)
    for o <- opts do
      oldSum += BigDecimal(o.score.toDouble)
      o.score = roundToTwoDigits(o.score.toDouble - optionScore / opts.size)
      sum += BigDecimal(o.score.toDouble)
    (oldSum - BigDecimal(optionScore) - sum)
      .round(new java.math.MathContext(1))
      .toDouble

  private def roundToTwoDigits(score: Double): Double = math.round(score * 100) / 100.0

  override def compare(o: ExamSectionQuestion): Int = sequenceNumber - o.sequenceNumber

  override def equals(o: Any): Boolean = o match
    case e: ExamSectionQuestion => this.examSection == e.examSection && this.question == e.question
    case _                      => false

  override def hashCode: Int =
    val p = if examSection != null then examSection.hashCode else 0
    31 * p + (if question != null then question.hashCode else 0)
